pwnlib/tubes/remote.py

```diff
@@ -120,10 +119,16 @@ def _connect(self, fam, typ):
                 try:
                     sock.connect(sockaddr)
                     return sock
-                except socks.ProxyError:
-                    raise
-                except socket.error:
-                    pass
+                except IOError as e:
+                    if not isinstance(e, socket.error):
+                        raise
```

Comment:
This appears to render the following code useless, since a ProxyError is not a socket.error?

Reply:
This is a bit tricky, so I am not sure I got it right:
- We want to reraise all `socks.ProxyError`s.  We also want to silence all `socket.error`s (not sure why, but w/e, this is what the code has been doing so far).
- On py2 [`socket.error`](https://docs.python.org/2/library/socket.html#socket.error) is a child class of [`IOError`](https://docs.python.org/2/library/exceptions.html#exceptions.IOError), which is also the base class for `socks.ProxyError`. Therefore the following code is useless indeed.
- On py3 [`socket.error`](https://docs.python.org/3/library/socket.html#socket.error) is a deprecated alias for [`OSError`](https://docs.python.org/3/library/exceptions.html#OSError), which is also aliased as [`IOError`](https://docs.python.org/3/library/exceptions.html#IOError), and is therefore the base class for `socks.ProxyError`. The first check will always fail (so not sure whether we want to keep it), so the second check for `socks.ProxyError` is still needed, although I think we could also get away without importing `socks`, but through inspecting the exception's [`__mro__`](https://docs.python.org/3/library/stdtypes.html#class.__mro__) for classes matching `c.__module__ == "socks" and c.__name__ == "ProxyError"`.  This is not an expensive import however, and an error path anyway, so I think we can afford some extra time spent during actually importing the module.

Another idea I have is to cache the last exception anyway, and raise it just as [`socket.create_connection`](https://github.com/python/cpython/blob/2.7/Lib/socket.py#L541) does.  And I think this will be the best way in the end.